storage/postgres/labels.go

```diff
@@ -0,0 +1,210 @@
+				}
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func addLabel(ctx context.Context, newLabelFunc func(labelID string, labelKey string, labelValue string) Labelable, db pgDB, key string, values ...string) error {
+	for _, labelValue := range values {
+		uuids, err := uuid.NewV4()
+		if err != nil {
+			return fmt.Errorf("could not generate id for new label: %v", err)
+		}
+		labelID := uuids.String()
+		newLabel := newLabelFunc(labelID, key, labelValue)
+		labelTable, _, _ := newLabel.Label()
+		if _, err := create(ctx, db, labelTable, newLabel); err != nil {
+			if err == util.ErrAlreadyExistsInStorage {
+				return &query.LabelChangeError{Message: fmt.Sprintf("label with key %s and value %s already exists for this entity", key, labelValue)}
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func removeLabel(ctx context.Context, execer sqlx.ExtContext, labelable Labelable, referenceID, labelKey string, labelValues ...string) error {
+	labelTableName, referenceColumnName, _ := labelable.Label()
+	baseQuery := fmt.Sprintf("DELETE FROM %s WHERE key=? AND %s=?", labelTableName, referenceColumnName)
+	labelValuesCount := len(labelValues)
+	// remove all labels with this key
+	if labelValuesCount == 0 {
+		return executeNew(ctx, execer, baseQuery, []interface{}{labelKey, referenceID})
+	}
+	args := []interface{}{labelKey, referenceID}
+	if labelValuesCount == 1 {
+		args = append(args, labelValues[0])
+	} else {
+		args = append(args, labelValues)
+	}
+	// remove labels with a specific key and a value which is in the provided list
+	baseQuery += " AND val IN (?)"
+	sqlQuery, queryParams, err := sqlx.In(baseQuery, args...)
+	if err != nil {
+		return err
+	}
+	return executeNew(ctx, execer, sqlQuery, queryParams)
+}
+
+func buildQueryWithParams(extContext sqlx.ExtContext, sqlQuery string, baseTableName string, labelable Labelable, criteria []query.Criterion) (string, []interface{}, error) {
+	if len(criteria) == 0 {
+		return sqlQuery + ";", nil, nil
+	}
+
+	var queryParams []interface{}
+	var fieldQueries []string
+	var labelQueries []string
+
+	labelCriteria, fieldCriteria := splitCriteriaByType(criteria)
+
+	if len(labelCriteria) > 0 {
+		labelTableName, referenceColumnName, _ := labelable.Label()
+		labelSubQuery := fmt.Sprintf("(SELECT * FROM %[1]s WHERE %[2]s IN (SELECT %[2]s FROM %[1]s WHERE ", labelTableName, referenceColumnName)
+		for _, option := range labelCriteria {
+			rightOpBindVar, rightOpQueryValue := buildRightOp(option)
+			sqlOperation := translateOperationToSQLEquivalent(option.Operator)
+			labelQueries = append(labelQueries, fmt.Sprintf("%[1]s.key = ? AND %[1]s.val %[2]s %s", labelTableName, sqlOperation, rightOpBindVar))
+			queryParams = append(queryParams, option.LeftOp, rightOpQueryValue)
+		}
+		labelSubQuery += strings.Join(labelQueries, " AND ")
+		labelSubQuery += "))"
+
+		sqlQuery = strings.Replace(sqlQuery, "LEFT JOIN", "JOIN "+labelSubQuery, 1)
```

Comment:
the replace is a bit strange

Reply:
I don't understand. Change from LEFT JOIN to JOIN and append the remaining query exactly 1 times.

Reply:
I understand the query but what I mean that reconstructing the query seems strange - it kind of breaks the flow of appending stuff by switching the join type - its not a problem just looks strange